### PR TITLE
EIP 165 implementation

### PR DESCRIPTION
**Pull request by saxenism**
Solves #290

### CODE REVIEW

File: src/modules/lib/LibInterfaceId.sol [hunk 1]
```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
Comment (0xNuggan) [inline src/modules/lib/LibInterfaceId.sol:1]: What is the benefit of having a big file storing all the interfaceIDs? Isn't it easier to generate them on the fly since any contract checking them will know against what to compare? This way we avoid having to maintain a list, since we are going to be adding modules over time...

**Comment by 0xNuggan** [comment]
Also:
whereTests

File: src/orchestrator/Orchestrator.sol [hunk 2]
```diff
@@ -300,6 +265,26 @@ contract Orchestrator is IOrchestrator, ModuleManager {
...
+        require(
+            _contractAddress.code.length != 0,
+            "Contract Address need to passed here"
+        );
+
+        (bool success, bytes memory data) = _contractAddress.call(
+            abi.encodeWithSignature("supportsInterface(bytes4)", _interfaceId)
+        );
+
+        if (success && abi.decode(data, (bool))) {
```
Comment (FHieser) [inline src/orchestrator/Orchestrator.sol]: Cant u just return success && abi.decode(data, (bool)) here?

File: test/e2e/SingleVoteGovernorE2E.t.sol [hunk 3]
```diff
@@ -60,8 +60,8 @@ contract SingleVoteGovernorE2E is E2eTest {
             orchestratorConfig
         );
 
-        RebasingFundingManager fundingManager =
-            RebasingFundingManager(address(orchestrator.fundingManager()));
+        // RebasingFundingManager fundingManager =
```
Comment (FHieser) [inline test/e2e/SingleVoteGovernorE2E.t.sol]: Outcommented

File: src/factories/ModuleFactory.sol [hunk 4]
```diff
@@ -125,6 +126,18 @@ contract ModuleFactory is IModuleFactory, Ownable2Step {
...
     }
 
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override(ERC165)
+        returns (bool)
+    {
+        bytes4 interfaceId_IModuleFactory = type(IModuleFactory).interfaceId;
```
Comment (FHieser) [inline src/factories/ModuleFactory.sol]: Cant you just do:
return interfaceId == type(IModuleFactory).interfaceId || ....

**Review by FHieser — COMMENTED** [review]
Okay I have a comment that you might not like 🙈:
Can we move these to the start of the contract directly under the contract declaration?
I know the supportsInterface function is a public view function, but they dont have anything to do with the functionality of the contracts themselves but more with the interface implementation itself

File: src/factories/ModuleFactory.sol (hunk 4 above)
Comment (saxenism) [reply src/factories/ModuleFactory.sol]: Done

File: src/modules/lib/LibInterfaceId.sol (hunk 1 above)
Comment (saxenism) [reply src/modules/lib/LibInterfaceId.sol:1]: My thought process was regarding the late dependency injection usecase. When creating a new contract, you might get an address that either you can accept as a particular type of module or ask additionally for it's interface ID in the metadata, which you can compare against all the interface IDs of the inverter network by making a call to this library. Without this library, you'll probably have to import a particular type of module into the module that receives the metadata to calculate and then compare the interface IDs.

If this sounds like gibberish, happy to delete the library as well.

Comment (saxenism) [reply src/modules/lib/LibInterfaceId.sol:1]: Also, you'll see this being used sparsely in the Orchestrator.sol file.

**Comment by saxenism** [comment]
> Also: whereTests

Added the unit tests.

File: src/factories/ModuleFactory.sol [hunk 5]
```diff
@@ -32,7 +33,16 @@ import {
...
-contract ModuleFactory is IModuleFactory, Ownable2Step {
+contract ModuleFactory is IModuleFactory, Ownable2Step, ERC165 {
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override(ERC165)
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
```
Comment (FHieser) [inline src/factories/ModuleFactory.sol]: Isnt here the IModuleFactory missing?

File: src/modules/fundingManager/RebasingFundingManager.sol [hunk 6]
```diff
@@ -28,11 +28,22 @@ import {IFundingManager} from "src/modules/fundingManager/IFundingManager.sol";
 
 contract RebasingFundingManager is
     IFundingManager,
-    IRebasingERC20,
```
Comment (FHieser) [inline src/modules/fundingManager/RebasingFundingManager.sol:31]: Why was this removed?

File: src/modules/base/Module.sol [hunk 7]
```diff
@@ -31,7 +32,24 @@ import {IAuthorizer} from "src/modules/authorizer/IAuthorizer.sol";
...
+{
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override(ERC165)
+        returns (bool)
+    {
+        bytes4 interfaceId_IModule = type(IModule).interfaceId;
+        return interfaceId == interfaceId_IModule
```
Comment (FHieser) [inline src/modules/base/Module.sol]: Why isn this just return interfaceId == type(IModule).interfaceId

File: src/orchestrator/Orchestrator.sol
Comment (FHieser) [inline src/orchestrator/Orchestrator.sol:1]: didnt we want to include a way to find a module here?

File: src/modules/fundingManager/bondingCurveFundingManager/formula/BancorFormula.sol [hunk 8]
```diff
@@ -9,12 +13,23 @@ import {Utils} from
...
-///         2. Updating the Solidity version
-///         3. The version-specific removal of the 'public' keyword from the constructor
-///         4. The adjustments in the contract import methodology
-contract BancorFormula is IBancorFormula, Utils {
+/// @notice The sole modification applied to these contracts involves the alteration of the Solidity
+///         version and the version-specific removal of the 'public' keyword from the constructor,
+///         coinciding with adjustments in the contract import methodology.
+contract BancorFormula is IBancorFormula, Utils, ERC165 {
+
+    function supportsInterface(bytes4 interfaceId)
```
Comment (0xNuggan) [inline src/modules/fundingManager/bondingCurveFundingManager/formula/BancorFormula.sol]: I would remove this since we didn't want to touch the Formula file.

File: src/modules/lib/LibInterfaceId.sol (hunk 1 above)
Comment (0xNuggan) [reply src/modules/lib/LibInterfaceId.sol:1]: Hmm I see. Seems like overhead to me, what do you think, @FHieser ? Also, I think I'd remove the use in the Orchestrator, since it will only validate the core interface types.

File: src/orchestrator/Orchestrator.sol
Comment (FHieser) [reply src/orchestrator/Orchestrator.sol:1]: The Problem I had with this after looking at it again, was that there was a possibility of multiple modules supporting a Interface. This would mean for finding all modules with a specific interface we would have to create a dynamic array of module addresses, as we dont know how many there really are. My initial thought with this comment was that it would be nice to have nice and easy way to access modules that way, but simulating a dynamic array is quite frankly a bit of a pain.
 As we currently already have a way to search for a module via name I chose not to implement this

**Review by FHieser — APPROVED** [review]
download

As the original Implementation was done by Rahul and I just did some finishing touches I will approve this myself xD
